File: sdk/java/src/gov/auth/peer_t.java
package us.gov.auth;

import us.gov.id.peer_t.h;


class peer_t extends id.peer_t {
        enum stage_t {
            denied=0,
            authorized,
            num_stages
        };
        constexpr static array<const char*,num_stages> stagestr={"denied","authorized"};

        virtual const keys& get_keys() const=0;

        virtual void verification_completed() override;

	virtual bool authorize(const pubkey_t& p) const=0;
 void dump(ostream& os) const;
                virtual void dump_all(ostream& os) const override {
                        dump(os);
                        b::dump_all(os);
                }

        //virtual string run_auth() override;

        peer_t(int sock=0);
        virtual ~peer_t();
        stage_t stage{denied};
};

}
}}

#endif

constexpr array<const char*,c::num_stages> c::stagestr;

c::peer_t(int sock):b(sock) {
}

c::~peer_t() {
}
 
void c::verification_completed() {
    b::verification_completed();
    if (!verification_is_fine()) {
        disconnect();
        return;
    }
   	if (!authorize(pubkey)) {
        disconnect();
        return;
   	}
   	stage=authorized;
}
void c::dump(ostream& os) const {
    os << this << ' ' << pubkey << "- " << stagestr[stage] << endl;
}

File: gov/engine/peer_t.cpp
#include "peer_t.h"
#include "daemon.h"
#include "protocol.h"
#include "networking.h"
#include <us/gov/likely.h>

using namespace us::gov::engine;
using namespace std;

constexpr array<const char*,peer_t::num_stages> peer_t::m_stagestr;

void peer_t::verification_completed() {
    b::verification_completed();

    if (unlikely(!verification_is_fine())) {
        cout << "disconnect peer, pubkey not verified." << endl;
        disconnect();
        return;
    }

    static_cast<networking*>(parent)->get_parent()->get_auth_app()->basic_auth_completed(this);

    if (unlikely(stage==sysop)) { 
        if (static_cast<networking*>(parent)->get_parent()->sysop_allowed) {
            us::gov::auth::peer_t::set_stage(us::gov::auth::peer_t::authorized);
            send(new datagram(us::gov::protocol::sysop,"go ahead"));
        }else {
            us::gov::auth::peer_t::set_stage(us::gov::auth::peer_t::denied);
        }
    }

    if (unlikely(us::gov::auth::peer_t::get_stage()!=us::gov::auth::peer_t::authorized)) {
        cout << "disconnect peer, not authorized." << endl;
        disconnect();
        return;
    }
}

void peer_t::dump(ostream& os) const {
    os << "engine: " << this << " stage: " << stagestr[stage] << endl;
}


const keys& peer_t::get_keys() const {
    assert(parent != 0);
    return static_cast<id::daemon*>(parent)->get_keys();
}

bool peer_t::authorize(const pubkey_t& p) const {
    if(unlikely(parent == 0)) return true;
    return not static_cast<id::daemon*>(parent)->is_duplicate(p);
}

